### PR TITLE
Add webknossos_annotation.py

### PR DESCRIPTION
**Pull request by jingjingwu1225**

### CODE REVIEW

File: scripts/webknossos_annotation.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+
+def find_borders(img):
+    if np.max(img) == 0:
+        return [-1, -1, -1, -1]
+    t = cal_distance(img)
+    b = cal_distance(img[::-1]) 
+    l = cal_distance(np.rot90(img, k=3)) 
+    r = cal_distance(np.rot90(img, k=1))
+
+    # return [max(0, k-2) for k in [t, b, l, r]]
```
Comment (kabilar) [inline scripts/webknossos_annotation.py:243]: Hi @jingjingwu1225, I would suggest removing all commented code.

```suggestion
```

File: scripts/webknossos_annotation.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+def contour_to_mask(mask, value):
+    h, w = mask.shape[:2]
+    kernel = np.ones((5, 5), np.uint8)
+    closed_mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel) 
+    # closed_mask = skimage.morphology.binary_closing(mask) 
+    # plt.figure()
+    # plt.subplot(1,3,1)
+    # plt.imshow(mask)
+    # plt.subplot(1,3,2)
+    # plt.imshow(closed_mask)
```
Comment (kabilar) [inline scripts/webknossos_annotation.py:256]:
```suggestion
```

File: scripts/webknossos_annotation.py [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+    single_mask = binary_fill_holes(closed_mask)
+
+    # contours = measure.find_contours(closed_mask) 
+    # single_mask = np.zeros_like(mask) 
+    # print(len(contours))
+
+    # for contour in contours:
+    #     cur_mask = polygon2mask((h, w), contour)  
+    #     single_mask |= cur_mask
```
Comment (kabilar) [inline scripts/webknossos_annotation.py:265]:
```suggestion
```

File: scripts/webknossos_annotation.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    }
+    print(opt)
+
+
+    nblevel = 9
+    # Write each level
+    for level in range(nblevel):
+        omz_res = omz_data[level]
+        size = omz_res.shape[-2:]
+        shape = [1, 20] + [i for i in size]
```
Comment (balbasty) [inline scripts/webknossos_annotation.py:115]: Same comment - do we need to harcode this?

File: scripts/webknossos_annotation.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+        compressor_opt = ast.literal_eval(compressor_opt)
+
+    # Prepare Zarr group
+    store = zarr.storage.DirectoryStore(out)
+    omz = zarr.group(store=store, overwrite=True)
+
+
+    dic_EB = {0:0, 1:1, 9:2, 2:3, 4:4, 10:5, 11:6, 8:7, 3:8}
+    dic_JS = {0:0, 4:2, 5:3, 9:4, 6:5, 2:6, 7:7, 3:8}
+    dic_JW = {0:0, 1:1, 2:2, 3:3, 4:4, 6:5, 7:6, 8:7, 9:8}
```
[the remark this answers is not in the document]
Comment (balbasty) [reply scripts/webknossos_annotation.py:96]: Also, should this be an input option somehow?

File: scripts/webknossos_annotation.py [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+                    for j in range(nj):
+                        print(f'\r{i+1}/{ni}, {j+1}/{nj}', end=' ')
+                        start_x, end_x = i*max_load, min((i+1)*max_load, subdat_size[-2])
+                        start_y, end_y = j*max_load, min((j+1)*max_load, subdat_size[-1])
+                        array[0, idx,  t + start_x: t + end_x, l + start_y: l + end_y] = subdat[start_x: end_x, start_y: end_y]
+                print('')
+                
+        
+    # Write OME-Zarr multiscale metadata
+    print('Write metadata')
```
Comment (balbasty) [inline scripts/webknossos_annotation.py:164]: Do the annotation have the exact same shape as the annotated zarr file? I feel like we should just copy that info from the input zarr.

File: scripts/webknossos_annotation.py [hunk 7]
```diff
@@ -0,0 +1,303 @@
...
+    kernel = np.ones((5, 5), np.uint8)
+    closed_mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel) 
+    single_mask = binary_fill_holes(closed_mask)
+
+    single_mask = np.where(single_mask, value, 0)
+    return single_mask
+
+
+def generate_mask(mask):
+    final_mask = np.zeros_like(mask).astype(np.uint8)
```
Comment (balbasty) [inline scripts/webknossos_annotation.py:257]: What is this doing? Is it a good thing to process the data in a "conversion" script?

**Review by kabilar — COMMENTED** [review]
Hi @jingjingwu1225, while the details of this script are fresh on our minds, perhaps this is a good time to add unit and/or integration tests alongside the script?  Integrating these tests within a GitHub Actions workflow would be straightforward and I can help with this process.  Thanks for considering.

cc @balbasty

**Comment by jingjingwu1225** [comment]
Hi @kabilar, thanks for the comment. I'll modify the script first based on comments of Yael and then we can start integrating tests, you help is appreciated!

**Comment by balbasty** [comment]
@kabilar @jingjingwu1225 We're about to merge the refactored package in main. It would be great to integrate the script and tests into this refactored package. @calvinchai can provide guidance.

**Pull request closed without merging**